plaso/parsers/recycler.py

```diff
@@ -39,178 +37,189 @@ def __init__(self):
-      parser_mediator.ProduceExtractionError(
-          'unsupported format version: {0:d}.'.format(
-              header_struct.format_version))
-      return
+    file_header_map = self._GetDataTypeMap('recycle_bin_metadata_file_header')
 
-    if header_struct.deletion_time == 0:
+    try:
+      file_header, _ = self._ReadStructureFromFileObject(
+          file_object, 0, file_header_map)
+    except (ValueError, errors.ParseError) as exception:
+      raise errors.UnableToParseFile((
+          'Unable to parse Windows Recycle.Bin metadata file header with '
+          'error: {0!s}').format(exception))
+
+    if file_header.format_version not in self._SUPPORTED_FORMAT_VERSIONS:
+      raise errors.UnableToParseFile(
+          'Unsupported format version: {0:d}.'.format(
+              file_header.format_version))
+
+    if file_header.deletion_time == 0:
       date_time = dfdatetime_semantic_time.SemanticTime('Not set')
     else:
       date_time = dfdatetime_filetime.Filetime(
-          timestamp=header_struct.deletion_time)
+          timestamp=file_header.deletion_time)
 
     event_data = WinRecycleBinEventData()
-    event_data.original_filename = self._ReadFilename(
-        parser_mediator, file_object, header_struct.format_version)
-    event_data.file_size = header_struct.file_size
+    try:
+      event_data.original_filename = self._ParseOriginalFilename(
+          file_object, file_header.format_version)
+    except (ValueError, errors.ParseError) as exception:
+      parser_mediator.ProduceExtractionError(
+          'unable to parse original filename with error: {0!s}.'.format(
+              exception))
+
+    event_data.file_size = file_header.original_file_size
 
     event = time_events.DateTimeValuesEvent(
         date_time, definitions.TIME_DESCRIPTION_DELETED)
     parser_mediator.ProduceEventWithEventData(event, event_data)
 
 
-class WinRecyclerInfo2Parser(interface.FileObjectParser):
+class WinRecyclerInfo2Parser(dtfabric_parser.DtFabricBaseParser):
   """Parses the Windows Recycler INFO2 file."""
 
   NAME = 'recycle_bin_info2'
   DESCRIPTION = 'Parser for Windows Recycler INFO2 files.'
 
-  _FILE_HEADER_STRUCT = construct.Struct(
-      'file_header',
-      construct.ULInt32('unknown1'),
-      construct.ULInt32('unknown2'),
-      construct.ULInt32('unknown3'),
-      construct.ULInt32('record_size'),
-      construct.ULInt32('unknown4'))
-
-  _RECYCLER_RECORD_STRUCT = construct.Struct(
-      'recycler_record',
-      construct.ULInt32('index'),
-      construct.ULInt32('drive_number'),
-      construct.ULInt64('deletion_time'),
-      construct.ULInt32('file_size'))
-
-  _ASCII_STRING = construct.CString('string')
+  _DEFINITION_FILE = 'recycler.yaml'
 
   _RECORD_INDEX_OFFSET = 0x104
   _UNICODE_FILENAME_OFFSET = 0x118
 
-  def _ParseRecord(
-      self, parser_mediator, file_object, record_offset, record_size):
-    """Parses an INFO-2 record.
+  def _ParseFileEntry(
```

Comment:
Can you rename this method and the file_entry* args, please? It's confusing, given there's also dfvfs file_entries all over the codebase.

Reply:
ack.